### PR TITLE
Pre-encoding scripts

### PR DESCRIPTION
**Pull request by williamberman**
see PR comments

### CODE REVIEW

File: scripts/pre_encode.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+#
+# We re-upload a single new set of tar archive containing all embeddings:
+# <n>.tar
+# - <n><x>.pth
+# - <n><x>.pth
+# - <n><x>.pth
+# - ...
+#
+# Re-uploading a single new tar archive simplifies the pre-embedding and training code.
+# The only downside is the downloading of an un-used embedding at train time, but this
```
Comment (pcuenca) [inline scripts/pre_encode.py:18]: Do you mean that we need either the f8 or f16 encoding, but not both? I think it should be fine.

**Comment by isamu-isozaki** [comment]
@williamberman nice! I remember @patil-suraj mentioning that he didn't want to pre-encode the images with the vqgan since he wanted to do random cropping. Though when I was checking the performance, I think this will save around 60.529/199.07=30% of the time I think since that's the amount I noticed being spent on the prepare_inputs_and_labels function(though this is with t5)

File: scripts/pre_encode.py (hunk 1 above)
Comment (isamu-isozaki) [reply scripts/pre_encode.py:18]: @pcuenca I think both for the 2 stages of muse

Comment (williamberman) [reply scripts/pre_encode.py:18]: Here meaning just we upload both encodings as a part of the same set of shards. An individual training run will only use one of the two embeddings depending on what's configured as the vae

Comment (pcuenca) [reply scripts/pre_encode.py:18]: Yes, that's what I meant, thanks!

File: training/data.py [hunk 2]
```diff
@@ -338,3 +338,89 @@ def eval_dataset(self):
...
     def eval_dataloader(self):
         return self._eval_dataloader
+
+
+class Text2ImagePreEncodedDataset:
+    def __init__(
+        self,
+        train_shards_path_or_url: Union[str, List[str]],
+        eval_shards_path_or_url: Union[str, List[str]],
+        vae_checkpoint: str,
```
Comment (patil-suraj) [inline training/data.py:348]: (nit)

We can use the `Text2ImageDataset` for this, with some additional arg for pre-encode logic.

File: training/train_muse.py [hunk 3]
```diff
@@ -231,25 +231,32 @@ def main():
...
     model = model_cls(**config.model.transformer)
     mask_id = model.config.mask_token_id
 
     # Freeze the text model and VQGAN
-    text_encoder.requires_grad_(False)
-    vq_model.requires_grad_(False)
+    if text_encoder is not None:
+        text_encoder.requires_grad_(False)
+    if vq_model is not None:
+        vq_model.requires_grad_(False)
```
Comment (patil-suraj) [inline training/train_muse.py:259]:
```suggestion
    if not config.model.pre_encoded:
        text_encoder.requires_grad_(False)
        vq_model.requires_grad_(False)
```

File: scripts/pre_encode.py [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+#
+# Shard counts:
+# COYO) 74,752 shards (0-74,751)
+# laion 6a) 1,211 shards (0 - 1,210)
+# laion 5a) 60,581 shards (0 - 60,580)
+#
+# Encoding times using 8 nodes:
+# COYO) 20h48m
+# laion 6a) 23.4 minutes
+# laion 5a) 19h43m
```
Comment (williamberman) [inline scripts/pre_encode.py:27]: high level

File: scripts/pre_encode.py [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+
+    For the list of tar writers, we keep at most 5 open at a time. When we need to
+    open an additional writer, we close the earliest opened one assuming that we have
+    finished writing to it as the archives are read sequentially. This is an assumption
+    but 5 is a safe buffer as we realistically will never be writing to more than 2 at a time
+    for a reasonably sized thread pool.
+
+    The list of tar writers is managed with a global lock because it opens a sub process and
+    iirc Popen is not thread safe. Additionally each tar writer is managed with its own lock
+    because writes are not thread safe and can corrupt the archive.
```
Comment (williamberman) [inline scripts/pre_encode.py:111]: implementation details

**Comment by isamu-isozaki** [comment]
LGTM! One q, are we also interested in pre-encoding t5 embeddings?

**Comment by williamberman** [comment]
> LGTM! One q, are we also interested in pre-encoding t5 embeddings?

Nope!

**Pull request merged**